### DISCUSSION TITLE
Separate track types and ride types?

### DISCUSSION THREAD

**Discussion by X123M3-256** [OpenRCT2/OpenRCT2#18392] [Feature requests]
Currently, the game has no distinction between ride types and track types. I think this ought to change so that track types have their own ID and descriptor which the ride type references. There's a few reasons I think this makes sense:

* Different ride types can use the same track style. Currently this is isn't an issue because the two ride types can share the same track paint code. But if we want to implement ride type objects in future, it wouldn't make sense if every Intamin ride type has to include a copy of the Intamin track sprites, it should be able to reference the existing Intamin track (or a custom track type object, which I think should really be separate from the ride type object)

* Some ride types should be able to use more than one track type (particularly the giga coaster). Currently, if you wanted to implement this you'd need multiple ride type IDs and corresponding RTDs for the same ride type - and many of the fields would be redundant. I think it would be better if each different track style was its own track type, and then the ride type descriptor can reference more than one.

Some ride types already have similar functionality already but it is currently done by making multiple versions of every track piece - this would be a much cleaner way to do it, and it also means you could easily add an "Allow arbitrary track type changes" cheat using the same UI (functionality that already exists but is currently unusable).

* Ride types can be flat rides, which don't have a track type at all. If, for example, we want to add a UI to select the track type per track piece, it would make sense not to list ride types that aren't tracked rides, or to list separately rides that use identical track sprites. This could be done by adding a ride type flag that indicates if it should show in that list, but I think it makes more sense if the type field in the TrackElement refers to a track type rather than a ride type.

* If ride types were separate from track types, then ride type objects would be much easier to implement than track type objects. I think that the ability to have ride types as object files would be useful even while the track sprites remain hardcoded, because the existing track types already cover a lot of cases. This would, for example, allow the Vekoma looper to be implemented as a custom object using the LIM track sprites but without the LIM coaster's high intensity rating (and making chain lifts and zero G rolls available by default).

It is worth mentioning that vanilla .DAT files had a few features that were removed in OpenRCT2, such as the ability to display in the build menu as as a separate ride type rather than an alternate vehicle type, and the ability to change available track pieces and max build heights. This would add these capabilities back in a cleaner way, and it'd be a lot easier to do than making track types into object files (even though I really want that to happen eventually).

**Comment by karst** [comment]
This would also be incredibly helpful for changing arbitrary track types during construction in game, like a drop down menu in the construction. This would also make things possible to have covered pieces available as alternative track type rather than being a new element. Would add possibilities to have different spines for coaster types.

![](https://cdn.discordapp.com/attachments/220130899038175232/1033382119650635826/PXL_20221016_151818559.jpg)
Mack Spine

![](https://cdn.discordapp.com/attachments/220130899038175232/1033382380859301988/PXL_20220831_174415750.jpg)
Intamin Spine

I would also like to mention it is probably worth merging regular and inverted pieces in the same track code. Inverted coasters are I believe only 6 px higher than their flying counter part and that's across the board, this way you could make it so someone could make a (not so great looking but still something) zero-g stall, and whatnot. Turn all inversions into flying inversions so they can be used either way (when using cheats), etc. Would make everything overall cleaner and easier to use.

**Comment by spacek531** [comment]
~~It was my understanding this was the grand plan - to have separate objects for track paint and ride types, and the "TrackType" field in TrackElement will be updated to refer to track type objects instead of ride type objects.~~ It seems you are proposing something slightly different - giving each *paint code* its own identifier instead of referring to it tangentially through the ride type? This seems like a temporary solution to the technological debt of ride type and track type objects, but perhaps it does have merit in that service?

**Comment by karst** [comment]
I wanted to work out some ideas on Track Type objects and how to get the implemented. 

_The following plan is WIP_

To start we **need** to unite all track drawing. Here's a plan on how we could go about this:
- Rework TrackPaint.cpp to be readable like the other track code.
- Unite all ride types as flying rides. With this make all inversions flying inversions.
- Generalize support types in one file the json can redirect to. It also needs to allow to raise and lower and select the correct x/y/z for shape, placement and height.
- G1 needs to be resorted in one way or another to fit with order of other rides. It needs to make it so everything can be done with a simple 0 to 3 (direction). This should not be hardcoded in G1, but rather have a document that can convert this specifically for G1. People making new track will need to follow the standard of 0 to 3 for each track element. If multiple sprites are on a case we can add an "Extra" sprite, but those would have to be after the initial 0 to 3, to keep the maths at ease.

Json tags for track objects:
General:
- supportType (to allow selecting what supports it uses, would also allow for hotswapping supports using a cheat)
- isInvertedRide (This might actually fit better on a ride type, it would raise the track by 6px)
- fallbackId (if a ride isn't standalone, but uses elements from another ride, think of the LIM coaster, or Classic woodie)

Track Specific
- isInverted
- isCablelift
- isChainlift
(- isBooster)
(- isBrake)
(- isBlockbrake)

As a bonus I had the idea to separate trackitecture and add them to the scenery menu on a tab on its own. See #19710 for details.

**Reply by duncanspumpkin** [reply]
I really don't see any reason to sort g1 or g2. They are just offsets the order shouldn't matter. You should just refer to tables that have the offsets rather than the current assumptions in the codebase that assume that the next sprite will be x.

**Reply by X123M3-256** [reply]
> I really don't see any reason to sort g1 or g2. They are just offsets the order shouldn't matter.

The order matters if you want to eventually convert track types to object files, which I think we should. If the sprites are all in a standard order like the vehicle sprites are then there's no need for objects to store the full list of sprite indices, only a set of flags to indicate which sprites are included. 

G2 track sprites, at least the ones that I added, should be in a consistent order, but it's not really the most logical order, it's more or less the order that I implemented the pieces. Would make sense to have inversion pieces grouped together etc.

**Reply by karst** [reply]
> I really don't see any reason to sort g1 or g2. They are just offsets the order shouldn't matter. 

I don't mind the order of track elements, but within the track elements the order should be the same. Which goes with NESW, NWSE, SWNE, SENW (or whatever default order is), then the extras in the same order afterwards.
Currently if you for example look at the steeplechase. It doesn't go 0 1 2 3, it goes like 0 6 12 18, which is just not great if most orders are 0 1 2 3. Sure it's still a sequence, but it's not great if all sequences are different. Therefore we need to make a list per track element to see if the sequences are consistent, and if not, what the sequence is. (In this case *6). I guess what I mean with re-ordering is listing the elements where the sequence is inconsistent with the default sequence. With some pieces it even goes back on itself (2 3 0 1), which also needs to be listed in an array so it can be read out in the correct order.

**Reply by spacek531** [reply]
All this reordering can be done in the object's image list

**Comment by J-PIE-314** [comment]
This request is similar to https://github.com/OpenRCT2/OpenRCT2/discussions/17583 and is related to https://github.com/OpenRCT2/OpenRCT2/discussions/17572.
